### ISSUE TITLE
BUG: broken comparison between ndarray subclass (unyt) and strings

### ISSUE THREAD

**Issue by neutrinoceros** [numpy/numpy#22744]
### Describe the issue:

19ba7c4529a17a10abf0d52615011336a0deb606 ([#22707](https://github.com/numpy/numpy/pull/22707)) broke string comparison between `unyt.unyt_quantity` object (fancy 0d numpy arrays with unit data), and strings.

The following minimal example used to be fine, but it crashes on the main branch of `numpy`

### Reproduce the code example:

```python
import unyt as un

a = 1 * un.cm
assert a != "hello"
```

### Error message:

```shell
-traceback
Traceback (most recent call last):
  File "/Users/robcleme/dev/numpy/t.py", line 4, in <module>
    assert a != "hello"
           ^^^^^^^^^^^^
  File "/Users/robcleme/.pyenv/versions/numpy-dev/lib/python3.11/site-packages/unyt/array.py", line 1850, in __array_ufunc__
    new_dtype = np.dtype("f" + str(inp1.dtype.itemsize))
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: data type 'f20' not understood
```

### NumPy/Python version information:

Python 3.11.0

### Context for the issue:

This change of behaviour breaks `yt`, see https://github.com/yt-project/yt/issues/4241
~I note that we haven't captured this regression via `unyt`'s on CI, so I'll add a test there.~ (edit: this is https://github.com/yt-project/unyt/pull/333)

It is unclear to me why this is happening, and I am not sure which of `numpy` or `unyt` this bug should be fixed in.

Ping @ngoldbaum for information

**Comment by seberg** [comment]
Hmmmpf, NumPy ideally would have given a `DeprecationWarning` here, but didn't because for 0-D arrays ("scalars"), it assumed that an error in the ufunc machinery must be due to incomparable input.
`unyt` breaks the assumption here, it raises an unrelated *real* error.

Now, of course what I did now to fix the whole thing is specific enough to fix the original `DeprecationWarning` for this corner case.

I can also see that for unyt it is a bit tricker since this affects the "scalar" unyt and it happened to work out correctly before!  But I do wonder if it isn't worse to deal with having the DeprecationWarning longer, rather than just dealing with the fallout here.

**Comment by seberg** [comment]
Worst case, you need to override `==` and `!=` yourself to catch the specific error, this is what NumPy is doing now.  The problem was that NumPy was catching unspecific errors.

There is an assumption of numeric types in the code, which doesn't work for `==` and `!=` which now work for incomparable types.  It is probably better to use `np.result_type()` but that won't solve the underlying issue.

What I did in NumPy now (and why it is fixed!), is that if that `np.result_type` (or similar) fails there, I catch the specific exception and return an all False/True array.

**Comment by seberg** [comment]
@mhvk ping just in case you have a very quick insight in how astropy units is dealing with this.

**Comment by mhvk** [comment]
I'd think this is a `unyt` problem: it seems that in its `__eq__` it is trying to parse the string as a unit or quantity, and is not guarding properly against failure.

In astropy, we have no problems with units/quantity - the only fix I had to make was for our table class, which involved a test that previously checked for the error due to the deprecation warning (which now longer is raised, so things just work).

**Comment by seberg** [comment]
I see there is some fix in unyt, thanks!  Should we close this, or is there need to discuss it more?

**Comment by neutrinoceros** [comment]
I'm working on releasing the patch in unyt right now. I intend to do a final check with yt, where the bug was originally detected, and I'll close this. Should be done in a couple hours tops

**Comment by neutrinoceros** [comment]
I confirm that the issue is resolved with the new release (unyt 2.9.3) thanks everyone here for your quick replies !
